File: packages/ferric/test/conformance/runtime/sync/byte-lexemes.test.ts
import { test } from "node:test";
import * as assert from "node:assert/strict";
import {
  Engine, EngineHandle, EnginePool, FerricStringBytes, FerricSymbolBytes,
  FerricInstanceName, Format, HaltReason, toWire, fromWire,
} from "../../../helpers/ferric";

const bytes = Uint8Array.from([0x61, 0, 0xff]);
const source = "(defrule emit (bytes ?x) => (printout t ?x))";
const constructors = { FerricStringBytes, FerricSymbolBytes, FerricInstanceName };

function typedValues() {
  return [new FerricStringBytes(bytes), new FerricSymbolBytes(bytes), new FerricInstanceName(bytes)];
}

function assertTypedValues(values: readonly unknown[]) {
  for (const [index, ctor] of Object.values(constructors).entries()) {
    const value = values[index];
    assert.ok(value instanceof ctor);
    assert.deepEqual(value.bytes, bytes);
    assert.throws(() => value.value, /UTF-8|utf-8/);
  }
}

test("native byte lexemes, checked output and all snapshot codecs", () => {
  const engine = new Engine();
  try {
    engine.assertFact("typed", ...typedValues());
    assertTypedValues(engine.findFacts("typed")[0].fields);
    for (const format of [Format.Bincode, Format.Json, Format.Cbor, Format.MessagePack, Format.Postcard]) {
      const restored = Engine.fromSnapshot(engine.serialize(format), format);
      try { assertTypedValues(restored.findFacts("typed")[0].fields); }
      finally { restored.close(); }
    }
    engine.load(source);
    engine.assertFact("bytes", new FerricStringBytes(bytes));
    assert.equal(engine.run().rulesFired, 1);
    assert.deepEqual(engine.getOutputBytes("t"), bytes);
    assert.throws(() => engine.getOutput("t"), /UTF-8|utf-8/);
  } finally { engine.close(); }
});

test("byte wrappers copy inputs and wire preserves nested byte payloads", () => {
  const input = Uint8Array.from(bytes);
  const value = new FerricStringBytes(input);
  input[0] = 0;
  assert.deepEqual(value.bytes, bytes);
  const copy = value.bytes;
  copy[0] = 0;
  assert.deepEqual(value.bytes, bytes);
  const wire = structuredClone(toWire({ fields: typedValues(), output: bytes }));
  const result = fromWire(wire, undefined, constructors) as { fields: unknown[]; output: Uint8Array };
  assertTypedValues(result.fields);
  assert.deepEqual(result.output, bytes);
});

test("worker byte facts, exact output and checked text survive transport", async () => {
  const handle = await EngineHandle.create({ source });
  try {
    await handle.assertFact("typed", ...typedValues());
    assertTypedValues((await handle.findFacts("typed"))[0].fields);
    await handle.assertFact("bytes", new FerricStringBytes(bytes));
    await handle.run();
    assert.deepEqual(await handle.getOutputBytes("t"), bytes);
    await assert.rejects(handle.getOutput("t"), /UTF-8|utf-8/);
  } finally { await handle.close(); }
});

test("pool returns raw output without replacement text and rehydrates typed facts", async () => {
  const pool = await EnginePool.create([{ name: "bytes", source }], { threads: 1 });
  try {
    const result = await pool.evaluate("bytes", { facts: [
      { kind: "ordered", relation: "typed", fields: typedValues() },
      { kind: "ordered", relation: "bytes", fields: [new FerricStringBytes(bytes)] },
    ] });
    assertTypedValues(result.facts.find((fact) => fact.relation === "typed")!.fields);
    assert.deepEqual(result.outputBytes.stdout, bytes);
    assert.equal(result.output.stdout, undefined);
    await pool.do("bytes", async (proxy) => {
      await proxy.reset();
      await proxy.assertFact("bytes", new FerricStringBytes(bytes));
      await proxy.run();
      assert.deepEqual(await proxy.getOutputBytes("t"), bytes);
      await assert.rejects(proxy.getOutput("t"), /UTF-8|utf-8/);
    });
  } finally { await pool.close(); }
});


for (const operation of ["type", "named"]) {
  test(`missing-instance ${operation} preserves typed facts and halts later actions`, () => {
    const engine = new Engine();
    try {
      engine.load(`(defgeneric named)
        (defmethod named ((?x INSTANCE-NAME)) unreachable)
        (defrule check (name ?x) =>
          (printout t (instance-namep ?x) crlf) (assert (before ?x))
          (${operation} ?x) (assert (after)))`);
      engine.reset();
      const payload = new TextEncoder().encode("missing");
      engine.assertFact("name", new FerricInstanceName(payload));
      const result = engine.run();
      assert.equal(result.haltReason, HaltReason.ActionError);
      assert.equal(result.rulesFired, 1);
      assert.ok(engine.diagnostics.length > 0);
      assert.equal(engine.getOutput("t"), "TRUE\n");
      const fields = engine.findFacts("before")[0].fields;
      assert.equal(fields.length, 1);
      assert.ok(fields[0] instanceof FerricInstanceName);
      assert.deepEqual(fields[0].bytes, payload);
      assert.deepEqual(engine.findFacts("after"), []);
    } finally { engine.close(); }
  });
}

File: bindings/go/byte_lexemes_test.go
package ferric

import (
	"context"
	"encoding/json"
	"reflect"
	"testing"
)

func TestByteLexemesAndAllSnapshotFormats(t *testing.T) {
	for _, format := range allFormats() {
		t.Run(format.name, func(t *testing.T) {
			lockThread(t)
			engine, err := NewEngine()
			if err != nil {
				t.Fatal(err)
			}
			defer mustClose(t, engine)
			values := []any{StringBytes("a\x00\xff"), SymbolBytes("s\x00\xff"), InstanceName("n\x00\xff")}
			id, err := engine.AssertFact("bytes", values...)
			if err != nil {
				t.Fatal(err)
			}
			fact, err := engine.GetFact(id)
			if err != nil {
				t.Fatal(err)
			}
			if !reflect.DeepEqual(fact.Fields, values) {
				t.Fatalf("fields = %#v", fact.Fields)
			}
			data, err := engine.Serialize(format.format)
			if err != nil {
				t.Fatal(err)
			}
			restored, err := NewEngine(WithSnapshot(data, format.format))
			if err != nil {
				t.Fatal(err)
			}
			defer mustClose(t, restored)
			facts, err := restored.FindFacts("bytes")
			if err != nil {
				t.Fatal(err)
			}
			if len(facts) != 1 || !reflect.DeepEqual(facts[0].Fields, values) {
				t.Fatalf("restored = %#v", facts)
			}
		})
	}
}

func TestByteWireJSONAndOutputAreLossless(t *testing.T) {
	values := []any{StringBytes("a\x00\xff"), SymbolBytes("s\xff"), InstanceName("n\xff"), []any{StringBytes("")}}
	for _, value := range values {
		wire, err := NativeToWireValue(value)
		if err != nil {
			t.Fatal(err)
		}
		data, err := json.Marshal(wire)
		if err != nil {
			t.Fatal(err)
		}
		var decoded WireValue
		if err := json.Unmarshal(data, &decoded); err != nil {
			t.Fatal(err)
		}
		actual, err := WireToNativeValue(decoded)
		if err != nil {
			t.Fatal(err)
		}
		if !reflect.DeepEqual(actual, value) {
			t.Fatalf("%#v != %#v (%s)", actual, value, data)
		}
	}
	original := EvaluateResult{Output: map[string]string{"stdout": "a\x00\xff", "stderr": "plain"}}
	data, err := json.Marshal(original)
	if err != nil {
		t.Fatal(err)
	}
	var decoded EvaluateResult
	if err := json.Unmarshal(data, &decoded); err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(decoded.Output, original.Output) {
		t.Fatalf("output changed: %q", decoded.Output)
	}
}

func TestOutputCopyPreservesInvalidUTF8(t *testing.T) {
	lockThread(t)
	engine, err := NewEngine(WithSource("(defrule emit (bytes ?x) => (printout t ?x))"))
	if err != nil {
		t.Fatal(err)
	}
	defer mustClose(t, engine)
	if _, err := engine.AssertFact("bytes", StringBytes("a\x00\xff")); err != nil {
		t.Fatal(err)
	}
	if _, err := engine.Run(context.Background()); err != nil {
		t.Fatal(err)
	}
	output, exists, err := engine.GetOutputE("t")
	if err != nil || !exists || output != "a\x00\xff" {
		t.Fatalf("output = %q, %v, %v", output, exists, err)
	}
}

func TestMissingInstanceLookupPreservesTypedName(t *testing.T) {
	for _, operation := range []string{"type", "named"} {
		t.Run(operation, func(t *testing.T) {
			lockThread(t)
			source := `(defgeneric named)
(defmethod named ((?x INSTANCE-NAME)) unreachable)
(defrule check (name ?x) =>
 (printout t (instance-namep ?x) crlf) (assert (before ?x))
 (` + operation + ` ?x) (assert (after)))`
			engine, err := NewEngine(WithSource(source))
			if err != nil {
				t.Fatal(err)
			}
			defer mustClose(t, engine)
			name := InstanceName("missing")
			if _, err := engine.AssertFact("name", name); err != nil {
				t.Fatal(err)
			}
			result, err := engine.Run(context.Background())
			if err != nil {
				t.Fatal(err)
			}
			if result.HaltReason != HaltActionError || result.RulesFired != 1 || len(engine.Diagnostics()) == 0 {
				t.Fatalf("missing-instance result = %#v, diagnostics = %v", result, engine.Diagnostics())
			}
			output, exists, err := engine.GetOutputE("t")
			if err != nil || !exists || output != "TRUE\n" {
				t.Fatalf("predicate output = %q, %v, %v", output, exists, err)
			}
			before, err := engine.FindFacts("before")
			if err != nil || len(before) != 1 || !reflect.DeepEqual(before[0].Fields, []any{name}) {
				t.Fatalf("typed name capture = %#v, %v", before, err)
			}
			after, err := engine.FindFacts("after")
			if err != nil || len(after) != 0 {
				t.Fatalf("later action = %#v, %v", after, err)
			}
		})
	}
}
